File: 07_Trees/06_Subtree_of_Another_Tree.cpp
class Solution
{
public:
    bool isSame(TreeNode *node, TreeNode *target)
    {
        if (!node || !target)
            return node == target;
        if (node->val != target->val)
            return false;
        return isSame(node->left, target->left) && isSame(node->right, target->right);
    }

    bool isSubtree(TreeNode *root, TreeNode *subRoot)
    {
        if (!root)
            return false;
        if (isSame(root, subRoot))
            return true;
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }
};

File: 07_Trees/06_Subtree_of_Another_Tree.py
class Solution:

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        def isSame(node: Optional[TreeNode], target: Optional[TreeNode]):
            if node == None and target == None:
                return True
            if node == None or target == None:
                return False
            
            if node.val != target.val:
                return False
            
            return isSame(node.left, target.left) and isSame(node.right, target.right)
        
        if not root:
            return False

        if isSame(root, subRoot):
            return True
        
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)